### PR TITLE
Feature/dynamic blocks

### PR DESCRIPTION
**Pull request by sy238-unikent**
The PR adds support to the API for "Dynamic Blocks", where a dynamic block is one which:

- generates its block field values dynamically via code, external api requests, database calls etc, on each request
- AND / OR generates dynamic pages to match URLs UNDER those of the page on which it is embedded.

e.g. a "Profiles" block at /site/profiles which displays a list of staff from a database then allows viewing individual staff profiles on "virtual" pages at /site/profiles/<username>.

Dynamic blocks must:
-  set an attribute "dynamic" to 1 within their definition.json.
- implement a dynamic block class named after their definition name and version which extends App\Models\Definitions\Block
- implement one or both of the route() or filterData() methods as required.

Dynamic page URLs are only matched via the /api/v1/routes/resolve API call (used by renderers), not via other API requests which operate on page ids.

Dynamic blocks are currently only used within the ResolvesRoutes trait and the PageTransformer, but could easily be adapted to, for example, provide extra validation logic on saving pages, etc.

See the dynamic-blocks.md file in the root of the repo for more information.

### CODE REVIEW

File: app/Models/DynamicPage.php [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+class DynamicPage extends Page
+{
+	public $site = null;
+	public $parent = null;
+	public $revision = null;
+
+	/**
+	 * @var Page|null The "real" page which is a parent of this one.
+	 */
+	protected $dynamicParent = null;
```
Comment (cfc7-unikent) [inline app/Models/DynamicPage.php]: do we need $dynamicParent as well as parent?

File: app/Events/FilterResponseData.php [hunk 2]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+namespace App\Events;
+
```
Comment (cfc7-unikent) [inline app/Events/FilterResponseData.php]: get rid of for now

File: app/Models/Definitions/BaseDefinition.php [hunk 3]
```diff
@@ -22,6 +23,27 @@
 
     use HasAttributes, HidesAttributes, GuardsAttributes;
 
+	/**
+	 * @var Dispatcher
+	 */
+    protected $dispatcher = null;
```
Comment (cfc7-unikent) [inline app/Models/Definitions/BaseDefinition.php]: remove - no events... yet

File: app/Models/Definitions/BaseDefinition.php [hunk 4]
```diff
@@ -22,6 +23,27 @@
...
+
+	/**
+	 * Do any setup, etc, such as hooking into events via $this->dispatcher->listen()
+	 */
+    public function boot() {}
+
+	/**
+	 * Register any event listeners defined in this block's definition.json file
+	 */
+    public function registerListeners()
```
Comment (cfc7-unikent) [inline app/Models/Definitions/BaseDefinition.php]: remove event stuff

File: app/Models/Definitions/BaseDefinition.php [hunk 5]
```diff
@@ -323,13 +345,42 @@ public static function fromDefinition($json)
...
+			require_once $file_path;
+			$instance = new $class_name();
+		}
+		else {
+			$instance = new static();
+		}
     	$instance->forceFill($definition);
-
-    	return $instance;
+		$instance->dispatcher = app('events');
```
Comment (cfc7-unikent) [inline app/Models/Definitions/BaseDefinition.php]: no more events sam you overarchitecturer

File: app/Models/Definitions/Block.php [hunk 6]
```diff
@@ -11,6 +10,32 @@ class Block extends BaseDefinition
         'fields' => 'array',
 	];
 
+	/**
+	 * Optionally filter the data for this block.
+	 * @param array $block_data Array of data defining this instance of this block.
+	 * @return mixed Array of data.
+	 */
+	public function filterData($block_data){ return $block_data; }
```
Comment (cfc7-unikent) [inline app/Models/Definitions/Block.php]: update signature

**Comment by sy238-unikent** [comment]
References to events removed as requested, other changes also made.

**Pull request merged**